sklearn/utils/validation.py

```diff
@@ -966,3 +967,151 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
+      ``ValueError``. Otherwise all non-significant imaginary parts that may
+      remain are removed.
+
+    - that eigenvalues are not all negative. If this check fails, it raises a
+      ``ValueError``
+
+    - that there are no significant negative eigenvalues (with absolute value
+      more than 1e-10 and more than 1e-5 times the largest positive
+      eigenvalue). If this check fails, it raises a ``PSDSpectrumWarning``. All
+      negative eigenvalues (even smaller ones) are set to zero in all cases.
+
+    - that the eigenvalues are well conditioned. That means, that the
+      eigenvalues are all greater than the maximum eigenvalue divided by 1e12.
+      If this check fails and ``warn_on_zeros=True``, it raises a
+      ``PSDSpectrumWarning``. All the eigenvalues that are too small are then
+      set to zero.
+
+    Parameters
+    ----------
+    lambdas : array-like
+        Array of eigenvalues to check / fix.
+
+    warn_on_zeros : bool, optional (default=False)
+        When this is set to ``True``, a ``PSDSpectrumWarning`` will be raised
+        when there are extremely small eigenvalues. Otherwise no warning will
+        be raised. Note that in both cases, extremely small eigenvalues will be
+        set to zero.
+
+    Returns
+    -------
+    lambdas_fixed : array-like
+        A fixed validated copy of the array of eigenvalues.
+
+    Examples
+    --------
+    >>> check_psd_eigenvalues([1, 2])      # nominal case
+    ... # doctest: +NORMALIZE_WHITESPACE
+    array([1, 2])
+    >>> check_psd_eigenvalues((5, 5j))     # significant imag part
+    ... # doctest: +ELLIPSIS +NORMALIZE_WHITESPACE
+    Traceback (most recent call last):
+        ...
+    ValueError: There are significant imaginary parts in eigenvalues (1.000000
+        of the maximum real part). The matrix is maybe not PSD, or something
+        went wrong with the eigenvalues decomposition.
+    >>> check_psd_eigenvalues((5, 5e-5j))  # insignificant imag part
+    ... # doctest: +NORMALIZE_WHITESPACE
+    array([5., 0.])
+    >>> check_psd_eigenvalues((-5, -1))    # all negative
+    ... # doctest: +ELLIPSIS +NORMALIZE_WHITESPACE
+    Traceback (most recent call last):
+        ...
+    ValueError: All eigenvalues are negative (maximum is -1.000000). The matrix
+        is maybe not PSD, or something went wrong with the eigenvalues
+        decomposition.
+    >>> check_psd_eigenvalues((5, -1))     # significant negative
+    ... # doctest: +NORMALIZE_WHITESPACE
+    array([5, 0])
+    >>> check_psd_eigenvalues((5, -5e-5))  # insignificant negative
+    ... # doctest: +NORMALIZE_WHITESPACE
+    array([5., 0.])
+    >>> check_psd_eigenvalues((5, 4e-12))  # bad conditioning (too small)
+    ... # doctest: +NORMALIZE_WHITESPACE
+    array([5., 0.])
+
+    """
+    # is the provided array in double precision (float64) ?
+    is_double_precision = True
+    try:
+        is_double_precision = lambdas[0].dtype == np.dtype('f')
```

Comment:
Would that work?

```py
is_double_precision = True
if isinstance(lambdas, np.ndarray):
    is_double_precision = lambdas.dtype == np.dtype('f')
```

You shouldn't need to check `lambdas[0]`

Reply:
What is lambdas is a list of numpy elements ? Then the check would be skipped. I admit that this is a twisted case but... can happen no?

Reply:
you mean a list of numpy arrays with 1 elements? I don't think it should be supported. 99% of the use case for this function is to take its input from `eigh` which always returns a 1d array

Reply:
no I meant a list of numpy elements such as `np.int32(15)`. Elements have a dtype too.
But you're right this is too cumbersome. 

Actually, taking this into consideration, we could actually restrict it to numpy array inputs only. I'll push something like this.